File: Leetcode/island_perimeter/island_perimeter.rb
# @param {Integer[][]} grid
# @return {Integer}
def island_perimeter(grid)
  count = 0
  grid.each_with_index do |arr, idx|
    arr.each_with_index do |n, i|
      if n == 1
        count += 4
        if idx > 1
          count -= 1
          # require 'pry'; binding.pry
        end
        if arr[i-1] == 1
          count -= 1
          # require 'pry'; binding.pry
        end
        if arr[i+1] == 1
          count -= 1
          # require 'pry'; binding.pry
        end
        if grid[idx+1] && grid[idx+1][i] == 1
          count -= 1
          # require 'pry'; binding.pry
        end
        if idx > 0 && grid[idx - 1] == 1
          count -= 1
          # require 'pry'; binding.pry
        end
      end
    end
  end
  count
end

p island_perimeter([[0,1,0,0],[1,1,1,0],[0,1,0,0],[1,1,0,0]])

File: Leetcode/island_perimeter/island_perimeter.py
import pdb

class Solution(object):
    def islandPerimeter(self, grid):
        count = 0
        
        for index, arr in enumerate(grid):
            for indx, num in enumerate(arr):
                if num == 1:
                    count += 4
                    if indx + 1 < len(arr) and arr[indx + 1] == 1:
                        count -= 1
                    if indx - 1 > -1 and arr[indx - 1] == 1:
                        count -= 1
                    if index > 0 and grid[index - 1][indx] == 1:
                        count -= 1
                    if index + 1 < len(grid) and grid[index + 1][indx] == 1:
                        count -= 1
                
        return count
        

solution = Solution()
print(solution.islandPerimeter([[0,1,0,0],[1,1,1,0],[0,1,0,0],[1,1,0,0]]))
print(solution.islandPerimeter([[1]]))
print(solution.islandPerimeter([[1,0]]))
print(solution.islandPerimeter([[1],[1]]))